crates/brainfuck_prover/src/components/instruction/table.rs

```diff
@@ -83,11 +85,76 @@ impl InstructionTable {
     pub fn get_row(&self, row: &InstructionTableRow) -> Option<&InstructionTableRow> {
         self.table.iter().find(|r| *r == row)
     }
+
+    /// Retrieves the last row in the Instruction Table.
+    ///
+    /// # Returns
+    /// An `Option` containing a reference to the last [`InstructionTableRow`] in the table,
+    /// or `None` if the table is empty.
+    pub fn last_row(&self) -> Option<&InstructionTableRow> {
+        self.table.last()
+    }
+}
+
+impl From<Vec<Registers>> for InstructionTable {
+    fn from(registers: Vec<Registers>) -> Self {
+        // Create a mutable copy of the input `registers` to sort it.
+        let mut sorted_registers = registers;
+        // Sort `sorted_registers` by
+        // 1. `ip` (instruction pointer)
+        // 2. `clk` (clock cycle)
+        sorted_registers.sort_by(|a, b| a.ip.cmp(&b.ip).then_with(|| a.clk.cmp(&b.clk)));
+
+        // Initialize a new instruction table to store the sorted and processed rows.
+        let mut instruction_table = Self::new();
+
+        // Use `current_ip` to track the current instruction pointer and group elements.
+        let mut current_ip = None;
+
+        for register in &sorted_registers {
+            // Check if the current register marks the end of the program with:
+            // - `ci` = 0 and `ni` = 0
+            // If so, stop adding rows and break out of the loop.
+            if register.ci == BaseField::zero() && register.ni == BaseField::zero() {
+                break;
+            }
+
+            // If the instruction pointer (`ip`) changes:
+            // - Add an extra row to ensure each instruction appears `m + 1` times, where `m` is the
+            //   number of occurrences.
+            if Some(register.ip) != current_ip {
+                current_ip = Some(register.ip);
+                if let Some(last_row) = instruction_table.last_row() {
+                    instruction_table.add_row(InstructionTableRow {
+                        ip: last_row.ip,
+                        ci: last_row.ci,
+                        ni: last_row.ni,
+                    });
+                }
+            }
```

Comment:
Why would we add extra rows to the instruction table? It represents the instructions executed during the run, thus it most not be modified from the provided execution trace.

Reply:
Indeed we need to add the program to the instruction table, besides the registers from the execution trace

But, adding the program from the execution would only result in adding the executed program, which might differ from the actual program? (non-traversed branches, dead code)?